src/lib/Inference.hs

```diff
@@ -1943,21 +1944,28 @@ checkHandlerBodyTyArg b cont = do
       cont v
 
 checkInstanceParams
-  :: forall i o e.
+  :: forall i o e any.
      (EmitsInf o, SinkableE e, HoistableE e, RenameE e)
-  => [UType i]
+  => Nest (RolePiBinder CoreIR) o any
+  -> [UType i]
   -> (forall o'. (EmitsInf o', Ext o o') => [CType o'] -> InfererM i o' (e o'))
   -> InfererM i o (Abs (Nest (RolePiBinder CoreIR)) e o)
-checkInstanceParams params cont = go params []
+checkInstanceParams bsTop params cont = go bsTop params []
   where
-    go :: forall o'. (EmitsInf o', Ext o o') => [UType i] -> [CType o']
+    go :: forall o' any'. (EmitsInf o', Ext o o')
+       => Nest (RolePiBinder CoreIR) o' any'
+       -> [UType i] -> [CType o']
        -> InfererM i o' (Abs (Nest (RolePiBinder CoreIR)) e o')
-    go []    ptys = Abs Empty <$> cont (reverse ptys)
-    go (p:t) ptys = checkUTypeWithMissingDicts p \ds getParamType -> do
-      mergeAbs <$> introDicts (eSetToList ds) do
-        pty <- getParamType
-        ListE ptys' <- sinkM $ ListE ptys
-        go t (pty:ptys')
+    go Empty []    ptys = Abs Empty <$> cont (reverse ptys)
+    go (Nest (RolePiBinder b k _ _) bs) (p:t) ptys = do
+      checkParamWithMissingDicts k p \ds getParamType -> do
+        mergeAbs <$> introDicts (eSetToList ds) do
+          pty <- getParamType
+          bsAbs <- sinkM $ Abs b (EmptyAbs bs)
+          Abs bs' UnitE <- applyAbs bsAbs (SubstVal pty)
+          ListE ptys' <- sinkM $ ListE ptys
+          go bs' t (pty:ptys')
+    go _ _ _ = error "zip error"
```

Comment:
What are these changes changing?  What are we reading from the class definition now and why?

Reply:
They're changing because we can no longer assume that all class parameters have type `Type`. (Specifically, they might have now have type `Heap`.)